src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxResolvingClassLoader.java

```diff
@@ -65,33 +62,37 @@
     }
 
     @Override public URL getResource(String name) {
-        return parentResourceCache.getUnchecked(getParent()).getUnchecked(name).orNull();
+        ClassLoader parentLoader = getParent();
+        return load(parentResourceCache, name, parentLoader, () -> Optional.fromNullable(parentLoader.getResource(name))).orNull();
     }
 
-    private interface CacheFunction<T> {
-        T compute(ClassLoader parentLoader, String name);
+    // We cannot have the inner cache be a LoadingCache and just use .get(name), since then the values of the outer cache would strongly refer to the keys.
+    private static <T> T load(LoadingCache<ClassLoader, Cache<String, T>> cache, String name, ClassLoader parentLoader, Supplier<T> supplier) {
+        try {
+            return cache.getUnchecked(parentLoader).get(name, () -> {
+                Thread t = Thread.currentThread();
+                String origName = t.getName();
+                t.setName(origName + " loading " + name);
+                long start = System.nanoTime(); // http://stackoverflow.com/q/19052316/12916
+                try {
+                    return supplier.get();
+                } finally {
+                    t.setName(origName);
+                    long ms = (System.nanoTime() - start) / 1000000;
+                    if (ms > 1000) {
+                        LOGGER.log(Level.INFO, "took {0}ms to load/not load {1} from {2}", new Object[] {ms, name, parentLoader});
+                    }
+                }
+            });
+        } catch (ExecutionException x) {
+            throw new UncheckedExecutionException(x); // should not be possible anyway
+        }
     }
 
-    private static <T> LoadingCache<ClassLoader, LoadingCache<String, T>> makeParentCache(final CacheFunction<T> function) {
-        return CacheBuilder.newBuilder().weakKeys().build(new CacheLoader<ClassLoader,LoadingCache<String, T>>() {
-            @Override public LoadingCache<String, T> load(final ClassLoader parentLoader) {
-                return CacheBuilder.newBuilder()./* allow new plugins to be used, and clean up memory */expireAfterWrite(15, TimeUnit.MINUTES).build(new CacheLoader<String, T>() {
-                    @Override public T load(String name) {
-                        Thread t = Thread.currentThread();
-                        String origName = t.getName();
-                        t.setName(origName + " loading " + name);
-                        long start = System.nanoTime(); // http://stackoverflow.com/q/19052316/12916
-                        try {
-                            return function.compute(parentLoader, name);
```

Comment:
So if I understand correctly, the problem was that here we close over `parentLoader` in the inner `LoadingCache`, so `weakKeys` didn't do anything because the inner `LoadingCache` always had a strong reference to the parent loader? And in the new code, we instead pass a reference to the parent loader via a `Supplier` that is used to load the value if needed and is not retained anywhere?

Reply:
> the problem was …

Exactly. The use of inner classes vs. `static` (nested) classes is not significant—the inner `LoadingCache` really needed a reference to the `parentLoader` in order to function.

> in the new code, we instead pass a reference to the parent loader

Well `load` here is just a helper method to avoid code duplication between the two caches, so the `Supplier` is not interesting. The key difference is that we are now using a plain `Cache` for the inner cache, which does not need to hold on to a `CacheLoader`, so there is no reference chain from the singleton outer cache to a `ClassLoader`. Instead, the `ClassLoader` is held only via local variable from a `loadClass` or `getResource` method, and `Cache.get(K,                       Callable<? extends V>)` is used instead of `LoadingCache.get(K)`.